File: src/main/java/com/ponggame/model/AIController.java
package com.ponggame.model;

import javafx.scene.shape.Rectangle;
import java.util.Random;

public class AIController {
    private Rectangle paddle;
    private Ball ball;
    private boolean isHorizontal; // true = moves left/right, false = moves up/down
    private double minPos, maxPos; // Movement bounds

    private static final double MAX_SPEED = 4.0;
    private static final double ERROR_MARGIN = 30.0;
    private static final int REACTION_DELAY = 8;

    private int frameCounter = 0;
    private double targetPos;
    private Random random = new Random();

    public AIController(Rectangle paddle, Ball ball, boolean isHorizontal, double minPos, double maxPos) {
        this.paddle = paddle;
        this.ball = ball;
        this.isHorizontal = isHorizontal;
        this.minPos = minPos;
        this.maxPos = maxPos;
        this.targetPos = isHorizontal ? paddle.getX() : paddle.getY();
    }

    public void update() {
        frameCounter++;
        if (frameCounter >= REACTION_DELAY) {
            decideTarget();
            frameCounter = 0;
        }
        movePaddle();
    }

    private void decideTarget() {
        if (ball == null)
            return;

        double error = (random.nextDouble() - 0.5) * 2 * ERROR_MARGIN;

        if (isHorizontal) {
            // Horizontal paddle (top/bottom) - track ball X
            double paddleWidth = paddle.getWidth();
            targetPos = ball.getX() - paddleWidth / 2 + error;
        } else {
            // Vertical paddle (left/right) - track ball Y
            double paddleHeight = paddle.getHeight();
            targetPos = ball.getY() - paddleHeight / 2 + error;
        }

        // Clamp to bounds
        double paddleSize = isHorizontal ? paddle.getWidth() : paddle.getHeight();
        targetPos = Math.max(minPos, Math.min(maxPos - paddleSize, targetPos));
    }

    private void movePaddle() {
        double currentPos = isHorizontal ? paddle.getX() : paddle.getY();
        double diff = targetPos - currentPos;

        double move = Math.signum(diff) * Math.min(Math.abs(diff), MAX_SPEED);

        if (isHorizontal) {
            paddle.setX(currentPos + move);
        } else {
            paddle.setY(currentPos + move);
        }
    }
}

File: server/game/AIController.js
/**
 * AI Controller - Server-side AI paddle logic
 * Tracks ball movement and moves paddle to intercept
 */

class AIController {
    constructor(position, difficulty = 'medium') {
        this.position = position; // 'left', 'right', 'top', 'bottom'
        this.difficulty = difficulty;
        this.reactionDelay = this.getReactionDelay();
        this.frameCounter = 0;
        this.targetY = 400; // Center of arena
        this.targetX = 400;
        this.errorMargin = this.getErrorMargin();
    }

    getReactionDelay() {
        switch (this.difficulty) {
            case 'easy': return 15;
            case 'medium': return 8;
            case 'hard': return 3;
            default: return 8;
        }
    }

    getErrorMargin() {
        switch (this.difficulty) {
            case 'easy': return 60;
            case 'medium': return 30;
            case 'hard': return 10;
            default: return 30;
        }
    }

    getSpeed() {
        switch (this.difficulty) {
            case 'easy': return 4;
            case 'medium': return 6;
            case 'hard': return 8;
            default: return 6;
        }
    }

    update(paddle, ball, arenaSize) {
        this.frameCounter++;

        // Only update target every N frames (reaction delay)
        if (this.frameCounter >= this.reactionDelay) {
            this.frameCounter = 0;
            this.calculateTarget(paddle, ball, arenaSize);
        }

        // Move paddle toward target
        return this.moveTowardTarget(paddle, arenaSize);
    }

    calculateTarget(paddle, ball, arenaSize) {
        const isHorizontal = this.position === 'top' || this.position === 'bottom';
        const speed = this.getSpeed();

        if (isHorizontal) {
            // Horizontal paddle - track ball's X position
            const ballApproaching = (this.position === 'bottom' && ball.vy > 0) ||
                (this.position === 'top' && ball.vy < 0);

            if (ballApproaching) {
                // Predict where ball will be
                this.targetX = ball.x + (Math.random() - 0.5) * this.errorMargin;
            } else {
                // Return to center
                this.targetX = arenaSize / 2;
            }
        } else {
            // Vertical paddle - track ball's Y position
            const ballApproaching = (this.position === 'right' && ball.vx > 0) ||
                (this.position === 'left' && ball.vx < 0);

            if (ballApproaching) {
                // Predict where ball will be
                this.targetY = ball.y + (Math.random() - 0.5) * this.errorMargin;
            } else {
                // Return to center
                this.targetY = arenaSize / 2;
            }
        }
    }

    moveTowardTarget(paddle, arenaSize) {
        const speed = this.getSpeed();
        const isHorizontal = this.position === 'top' || this.position === 'bottom';
        let newPosition;

        if (isHorizontal) {
            const paddleCenter = paddle.x + paddle.width / 2;
            const diff = this.targetX - paddleCenter;

            if (Math.abs(diff) > speed) {
                newPosition = paddle.x + (diff > 0 ? speed : -speed);
            } else {
                newPosition = paddle.x;
            }

            // Clamp to arena bounds
            const minX = 30;
            const maxX = arenaSize - 30 - paddle.width;
            newPosition = Math.max(minX, Math.min(maxX, newPosition));

            return { x: newPosition };
        } else {
            const paddleCenter = paddle.y + paddle.height / 2;
            const diff = this.targetY - paddleCenter;

            if (Math.abs(diff) > speed) {
                newPosition = paddle.y + (diff > 0 ? speed : -speed);
            } else {
                newPosition = paddle.y;
            }

            // Clamp to arena bounds
            const minY = 30;
            const maxY = arenaSize - 30 - paddle.height;
            newPosition = Math.max(minY, Math.min(maxY, newPosition));

            return { y: newPosition };
        }
    }
}

module.exports = AIController;
